pkg/compact/compact.go

```diff
@@ -957,6 +957,16 @@ func (c *BucketCompactor) Compact(ctx context.Context) (rerr error) {
 
 		ignoreDirs := []string{}
 		for _, gr := range groups {
+			groupIDs := []string{}
+			for _, grID := range gr.IDs() {
+				groupIDs = append(groupIDs, grID.String())
+			}
+
+			groupDir := filepath.Join(c.compactDir, gr.Key())
```

Comment:
Yea, this is getting bit out of control. That's why we kind of were deleting things upfront on the compactor startup, always in the first place, to not care about every tiny individual case.

For example  now we will still miss case where we crash on left dir behind, but when we start some group is missing (maybe sharding configuration changed) and now we never clean that directory (:

Reply:
Should we consider rolling back #3031? :)

Reply:
Isn't that a bit of an extreme reaction? Yes, we were just cleaning everything without a second doubt but that's what I think should be improved & it's only natural that sometimes issues occur when doing a new thing. We aren't robots and sometimes mistakes slip through but that's why we have our release process and so on, right? If it's just a matter of cleaning the directories properly then I'm not convinced that the whole thing needs to be rolled back unless this becomes blocking on the Cortex part i.e. if you won't be able to update your dependency on Thanos because of this.

Give me a few days and I think we can fix it just fine.

Reply:
@bwplotka 
> For example now we will still miss case where we crash on left dir behind, but when we start some group is missing (maybe sharding configuration changed) and now we never clean that directory (:

How could this happen? The meta fetcher drops IDs that have been removed via the relabeling config and then that gets passed on to the syncer:
```go
		cf := baseMetaFetcher.NewMetaFetcher(
			extprom.WrapRegistererWithPrefix("thanos_", reg), []block.MetadataFilter{
				block.NewLabelShardedMetaFilter(relabelConfig),
				block.NewConsistencyDelayMetaFilter(logger, conf.consistencyDelay, extprom.WrapRegistererWithPrefix("thanos_", reg)),
				ignoreDeletionMarkFilter,
				duplicateBlocksFilter,
				noCompactMarkerFilter,
			}, []block.MetadataModifier{block.NewReplicaLabelRemover(logger, conf.dedupReplicaLabels)},
		)
sy, err = compact.NewSyncer(
			logger,
			reg,
			bkt,
			cf,
			duplicateBlocksFilter,
			ignoreDeletionMarkFilter,
			blocksMarked.WithLabelValues(metadata.DeletionMarkFilename),
			garbageCollectedBlocks,
			conf.blockSyncConcurrency)
```

And then we do this:
```go
		groups, err := c.grouper.Groups(c.sy.Metas())
		if err != nil {
			return errors.Wrap(err, "build compaction groups")
		}
```

So, I'm not sure how there could be compaction groups of blocks that have been filtered out. Could you please elaborate?

Reply:
> Isn't that a bit of an extreme reaction?

Sorry. Was a joke. I should have made it more clear, sorry.

Reply:
haha, yea let's iterate @GiedriusS  (:

Reply:
I mean both compaction and downsampling, but I just pointed one problem, there might be more. If there is nothing simple, then we need to just review properly etc, and I might need time for this. Just adding extra deletes "just in case" does not sound like efficient way of doing things. Not because it takes some compute power, is that it surprise reader and just unnecessary increase cognitive load.

Maybe I exaggerate, I will try to pull this code locally and check tomorrow (:

Reply:
@bwplotka so, did you have time? What's the verdict? Is there any other case I've missed?

Reply:
Had zero time, added to https://github.com/thanos-io/thanos/issues/3946

Reply:
Proposed something (: